Panic in backtick parsing

There is a panic in how backticks are parsed in config files.

Here is the panic/traceback:

```
panic: missing end quote

goroutine 1 [running]:
github.com/gravwell/gcfg.unquote(0xc0000ea300, 0x28, 0x5, 0xc0000ea300)
        /home/kris/mygo/pkg/mod/github.com/gravwell/gcfg@v1.2.7/read.go:45 +0x304
github.com/gravwell/gcfg.readIntoPass(0xc000087670, 0x715ea0, 0xc0000c2c60, 0xc0001365c0, 0xc0000b01e0, 0xc000160000, 0xd21, 0x1e00, 0xc0000ae500, 0x0, ...)
        /home/kris/mygo/pkg/mod/github.com/gravwell/gcfg@v1.2.7/read.go:160 +0xcd2
github.com/gravwell/gcfg.readInto(0x715ea0, 0xc0000c2c60, 0xc0001365c0, 0xc0000b01e0, 0xc000160000, 0xd21, 0x1e00, 0x0, 0x40c678)
        /home/kris/mygo/pkg/mod/github.com/gravwell/gcfg@v1.2.7/read.go:194 +0xda
github.com/gravwell/gcfg.ReadInto(0x715ea0, 0xc0000c2c60, 0x810100, 0xc0000a3200, 0xc00015a000, 0xd21)
        /home/kris/mygo/pkg/mod/github.com/gravwell/gcfg@v1.2.7/read.go:214 +0x132
github.com/gravwell/gcfg.ReadStringInto(...)
        /home/kris/mygo/pkg/mod/github.com/gravwell/gcfg@v1.2.7/read.go:221
github.com/gravwell/gravwell/v3/ingest/config.LoadConfigBytes(0x715ea0, 0xc0000c2c60, 0xc000158000, 0xd21, 0x1e00, 0x0, 0x0)
        /home/kris/githubwork/gravwell/ingest/config/loader.go:80 +0xd5
github.com/gravwell/gravwell/v3/ingest/config.LoadConfigFile(0x715ea0, 0xc0000c2c60, 0x7ab041, 0x23, 0x0, 0x0)
        /home/kris/githubwork/gravwell/ingest/config/loader.go:70 +0x306
main.GetConfig(0x7ab041, 0x23, 0x1400, 0x7efbfd4ba400, 0x20300000000000)
        /home/kris/githubwork/gravwell/ingesters/SimpleRelay/config.go:77 +0x76
main.main()
        /home/kris/githubwork/gravwell/ingesters/SimpleRelay/main.go:81 +0x65                                                                               
```


Here is the config snippet that triggered it:
```
[Listener "sd"]
        Bind-String="0.0.0.0:8888"
        Tag-Name=sdtest4
        Ignore-Timestamps=true
        Preprocessor=cvesplit
        Preprocessor=ts

[Preprocessor "cvesplit"]
        Type=jsonarraysplit
        Extraction="cves"
        Passthrough-Misses=false
        Force-JSON-Object=true
        Additional-Fields="ip,last_seen"

[Preprocessor "ts"]
        Type=regextimestamp
        Regex=`"last_seen":\s*"(?P<timestamp>[^"]+)"`
        TS-Match-Name=timestamp
```

--- a/scanner/scanner.go
+++ b/scanner/scanner.go
@@ -224,11 +224,14 @@ func (s *Scanner) scanValString() string {
 		// unquote in several places, so instead we REWRITE the
 		// string as a double quoted string with double escapes
 		// as we see them...
+		// We also escape any other double quotes that aren't already
+		// escaped.
 		var b strings.Builder
 		b.WriteRune('"')
 
 		s.next()
 		ch := s.ch
+		var escapeState bool
 		for ch != '`' {
 			if ch < 0 {
 				s.error(offs, "raw string not terminated")
@@ -236,8 +239,12 @@ func (s *Scanner) scanValString() string {
 			}
 
 			if ch == '\\' {
+				escapeState = !escapeState
+				b.WriteRune('\\')
+			} else if ch == '"' && !escapeState {
 				b.WriteRune('\\')
 			}
+
 			b.WriteRune(ch)
 
 			s.next()

--- a/scanner/scanner_test.go
+++ b/scanner/scanner_test.go
@@ -252,6 +252,66 @@ func TestScanValStringEOF(t *testing.T) {
 	}
 }
 
+func TestScanValStringQuote(t *testing.T) {
+	var s Scanner
+	src := "= \"value\""
+	f := fset.AddFile("src", fset.Base(), len(src))
+	s.Init(f, []byte(src), nil, 0)
+	s.Scan()              // =
+	_, _, lit := s.Scan() // value
+	if lit != `"value"` {
+		t.Errorf("bad literal: got %s, expected %s", lit, `"value"`)
+	}
+	if s.ErrorCount > 0 {
+		t.Error("scanning error")
+	}
+}
+
+func TestScanValStringBacktickWonkyQuotesDoubleEscaped(t *testing.T) {
+	var s Scanner
+	src := "= `v\"a\"\\\"lue`"
+	f := fset.AddFile("src", fset.Base(), len(src))
+	s.Init(f, []byte(src), nil, 0)
+	s.Scan()              // =
+	_, _, lit := s.Scan() // value
+	if lit != `"v\"a\"\\"lue"` {
+		t.Errorf("bad literal: got %s, expected %s", lit, `"v\"a\"\\"lue"`)
+	}
+	if s.ErrorCount > 0 {
+		t.Error("scanning error")
+	}
+}
+
+func TestScanValStringBacktickWonkyQuotes(t *testing.T) {
+	var s Scanner
+	src := "= `v\"a\"\"lue`"
+	f := fset.AddFile("src", fset.Base(), len(src))
+	s.Init(f, []byte(src), nil, 0)
+	s.Scan()              // =
+	_, _, lit := s.Scan() // value
+	if lit != `"v\"a\"\"lue"` {
+		t.Errorf("bad literal: got %s, expected %s", lit, `"v\"a\"\"lue"`)
+	}
+	if s.ErrorCount > 0 {
+		t.Error("scanning error")
+	}
+}
+
+func TestScanValStringBacktick(t *testing.T) {
+	var s Scanner
+	src := "= `value`"
+	f := fset.AddFile("src", fset.Base(), len(src))
+	s.Init(f, []byte(src), nil, 0)
+	s.Scan()              // =
+	_, _, lit := s.Scan() // value
+	if lit != `"value"` {
+		t.Errorf("bad literal: got %s, expected %s", lit, `"value"`)
+	}
+	if s.ErrorCount > 0 {
+		t.Error("scanning error")
+	}
+}
+
 // Verify that initializing the same scanner more then once works correctly.
 func TestInit(t *testing.T) {
 	var s Scanner

